trace/obfuscate.go

```diff
@@ -1,40 +1,204 @@
+// 3. It applies regex patterns to all obfuscate sensitive primitive string values.
+func (s *redactSensitiveInfoVisitor) EnterData(self interface{}, ctx vis.SpecVisitorContext, d *pb.Data) Cont {
+	if httpMeta := d.GetMeta().GetHttp(); httpMeta != nil {
+		var key string
+		switch httpMeta.Location.(type) {
+		case *pb.HTTPMeta_Auth:
+			return ObfuscatePrimitiveWithRedactedString(d)
+		case *pb.HTTPMeta_Cookie:
+			return ObfuscatePrimitiveWithRedactedString(d)
+		case *pb.HTTPMeta_Header:
+			header := httpMeta.GetHeader()
+			key = header.Key
+		case *pb.HTTPMeta_Query:
+			queryParam := httpMeta.GetQuery()
+			key = queryParam.Key
+		}
+		// Check if the key is in the list of keys to obfuscate.
+		if s.obfuscationOptions.SensitiveDataKeys.Contains(strings.ToLower(key)) {
+			return ObfuscatePrimitiveWithRedactedString(d)
+		}
+	}
+
+	primitive := d.GetPrimitive()
+	if primitive == nil {
+		// Not a primitive, regex will be applied to primitive string values only.
+		return Continue
+	}
+
+	stringValue := primitive.GetStringValue()
+	if stringValue == nil {
+		// Not a string, regex will be applied to string values only.
+		return Continue
+	}
+
+	for pattern, _ := range s.obfuscationOptions.SensitiveDataValuePatterns {
+		if pattern.MatchString(stringValue.Value) {
+			return ObfuscatePrimitiveWithRedactedString(d)
+		}
+	}
+
+	return Continue
+}
+
+func (s *redactSensitiveInfoVisitor) EnterHTTPBody(self interface{}, ctx vis.SpecVisitorContext, b *pb.HTTPBody) Cont {
+	node, _ := ctx.GetInnermostNode(reflect.TypeOf((*pb.Data)(nil)))
+	data := node.(*pb.Data)
+	if data == nil {
+		return Continue
+	}
+
+	// Traverse the data node and redact sensitive information.
+	s.traverseAndRedactSensitiveInfo(ctx, data)
+
+	return Continue
+}
+
+func (s *redactSensitiveInfoVisitor) traverseAndRedactSensitiveInfo(ctx vis.SpecVisitorContext, data *pb.Data) {
+	switch val := data.Value.(type) {
+	case *pb.Data_Struct:
+		// Traverse the struct's fields and redact sensitive information based on key.
+		structData := val.Struct
+		for fieldName, fieldData := range structData.GetFields() {
+			// If field Data is primitive or primitiveList, then check it's key and redact if needed
+			if spec_util.IsPrimitive(fieldData) || spec_util.IsPrimitiveList(fieldData) {
+				if s.obfuscationOptions.SensitiveDataKeys.Contains(strings.ToLower(fieldName)) {
+					ObfuscatePrimitiveWithRedactedString(fieldData)
+				}
+			} else {
+				// Recursively traverse the struct's fields.
+				childCtx := ctx.EnterStruct(structData, fieldName).(vis.SpecVisitorContext)
+				s.traverseAndRedactSensitiveInfo(childCtx, fieldData)
+			}
+		}
+	case *pb.Data_List:
+		listData := val.List
+		if items := listData.GetElems(); items != nil {
+			for i, itemData := range items {
+				// Step through each item in the list and traverse the data node.
+				// This list will never be primitive list, as primitive list is handled above.
```

Comment:
Don't understand this comment, sorry.

This doesn't seem correct, we can have a JSON list at the top level, or inside another list?

```
{
  "api_key" : [ [ "a", b", "c" ], ["d", "e", "f" ]
}
```